### PR TITLE
Feature/hedev 617 fees numeraire

### PR DESCRIPTION
**Pull request by 0xAplki**
## Description of Changes
- Conversion of input units to numeraire as discussed in previous discussions, improvement of #15 
- Added and tweaked assimilator tests plus helpers from #13 
- Updated `.gitignore`
- Updated `README.md`

[Link to Jira Ticket](https://halodao.atlassian.net/browse/HDEV-617)
[Link to Jira Ticket](https://halodao.atlassian.net/browse/HDEV-616)

## How To Test
`yarn clean && yarn compile-fix && yarn test-local`

### Developer Checklist:

- [ ] I have followed the guidelines in our Contributing document
- [ ] This PR has a corresponding JIRA ticket
- [ ] My branch conforms with our naming convention i.e. `feature/HDEV-XXX-description`
- [ ] I have written new tests for your core changes, as applicable
- [ ] I have successfully ran tests locally
- [ ] I have formatted my code using format document in VSCode

### Reviewers Checklist:

- [ ] Code is readable and understandable; any unclear parts have explanations
- [ ] I have successfully ran tests locally

### CODE REVIEW

File: contracts/core/CurveMath.sol [hunk 1]
```diff
@@ -101,22 +101,27 @@ library CurveMath {
         int128[] memory _nBals,
         int128 _inputAmt,
         uint256 _outputIndex
-    ) internal view returns (int128 outputAmt_) {
+    ) internal view returns (int128 outputAmt_, int128 accruedFees_) {
         outputAmt_ = -_inputAmt;
 
-        int128 _lambda = curve.lambda;
+        // int128 _lambda = curve.lambda;
```
Comment (schystz) [inline contracts/core/CurveMath.sol:107]: Any particular reason why remove this var and directly use `curve.lambda` on lines 123 and 124 below?

**Review by schystz — COMMENTED** [review]
@0xAplki looks good to me, I just a couple of minor suggestions and questions

File: contracts/FXPool.sol [hunk 2]
```diff
@@ -419,24 +433,35 @@ contract FXPool is IMinimalSwapInfoPool, BalancerPoolToken, Ownable, Storage, Re
...
+        assimilator_ = curve.assimilators[_derivative].addr;
+    }
+
     /// @notice view LP tokens and token needed for deposit
     function viewDeposit(bytes calldata userData) external view whenNotPaused returns (uint256, uint256[] memory) {
-        (uint256[] memory tokensIn, address[] memory assetAddresses) = abi.decode(userData, (uint256[], address[]));
-
-        uint256 totalDepositNumeraire = (_convertToNumeraire(tokensIn[0], _getAssetIndex(assetAddresses[0])) +
-            _convertToNumeraire(tokensIn[1], _getAssetIndex(assetAddresses[1]))) * 1e18;
+        (uint256 totalDepositNumeraire, ) = abi.decode(userData, (uint256, address[]));
```
[the remark this answers is not in the document]
Comment (0xAplki) [reply contracts/FXPool.sol:464]: thanks for catching!

File: contracts/core/CurveMath.sol [hunk 3]
```diff
@@ -101,22 +101,27 @@ library CurveMath {
...
         int128[] memory _weights = curve.weights;
 
         int128 _omega = calculateFee(_oGLiq, _oBals, curve, _weights);
         int128 _psi;
 
         for (uint256 i = 0; i < 32; i++) {
-            _psi = calculateFee(_nGLiq, _nBals, curve, _weights);
+            {
+                _psi = calculateFee(_nGLiq, _nBals, curve, _weights);
```
[the remark this answers is not in the document]
Comment (0xAplki) [reply contracts/core/CurveMath.sol:115]: to avoid stack too deep!

[the remark this answers is not in the document]
Comment (0xAplki) [reply contracts/core/CurveMath.sol:115]: you can have variable assigned outside then you can assign something inside that scope

File: contracts/core/CurveMath.sol (hunk 1 above)
Comment (0xAplki) [reply contracts/core/CurveMath.sol:107]: i just did a direct reference to `curve.lambda` to save some variables.  it's just used once anyway at line 123

File: contracts/core/CurveMath.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (schystz) [reply contracts/core/CurveMath.sol:115]: Oh, nice trick!

File: contracts/core/CurveMath.sol (hunk 1 above)
Comment (schystz) [reply contracts/core/CurveMath.sol:107]: Got it! Short answer: cause solidity sucks sometimes :D

File: contracts/core/FXSwaps.sol [hunk 4]
```diff
@@ -205,4 +214,16 @@ library FXSwaps {
     ) internal view returns (int128 amt_, int128 bal_) {
         return Assimilators.viewNumeraireAmountAndBalance(_assim, _amt, address(curve.vault), curve.poolId);
     }
+
+    // internal function to avoid stack too deep
+    function _calculateFeeInNumeraire(int128 _amtWithEpsilon, int128 _amtWithoutEpsilon)
```
Comment (ctverceles) [inline contracts/core/FXSwaps.sol:219]:
```suggestion
    function _calculateFeeInNumeraire(int128 _outputNumeraireAmt, int128 _inputNumeraireAmt)
```
@0xAplki better naming? 

`_amtWithoutEpsilon` to me doesn't really explain the nature of the variable. its true but it also does not contain lambda (its the input amount that the user gave). 

`_amtWithEpsilon` is actually not entirely accurate cause it also contains the result of the lambda calculation. and we're passing the output amount to the function param

File: contracts/core/FXSwaps.sol [hunk 5]
```diff
@@ -205,4 +214,16 @@ library FXSwaps {
...
+        internal
+        pure
+        returns (int128 feeInNumeraire)
+    {
+        // lambda + (amount w/ epsilon - amount without epsilon) in numeraire
+        // int128 _amtWithEpsilon = _amtWithoutEpsilon.us_mul(ONE + curve.epsilon);
+
+        return _amtWithEpsilon.sub(_amtWithoutEpsilon);
```
Comment (ctverceles) [inline contracts/core/FXSwaps.sol:227]:
```suggestion
        return _amtWithEpsilon.sub(_inputNumeraireAmt);
```
@0xAplki rename var to `_inputNumeraireAmt`

File: contracts/core/FXSwaps.sol [hunk 6]
```diff
@@ -205,4 +214,16 @@ library FXSwaps {
...
     }
+
+    // internal function to avoid stack too deep
+    function _calculateFeeInNumeraire(int128 _amtWithEpsilon, int128 _amtWithoutEpsilon)
+        internal
+        pure
+        returns (int128 feeInNumeraire)
+    {
+        // lambda + (amount w/ epsilon - amount without epsilon) in numeraire
+        // int128 _amtWithEpsilon = _amtWithoutEpsilon.us_mul(ONE + curve.epsilon);
```
Comment (ctverceles) [inline contracts/core/FXSwaps.sol:225]:
```suggestion
```
clean comments if all good @0xAplki ?

File: tests/local/pool.test.ts [hunk 7]
```diff
@@ -570,37 +416,18 @@ describe('FXPool', () => {
 
   // it('Previews swap caclculation from the onSwap hook using queryBatchSwap() ', async () => {})
   // it('Previews swap caclculation when providing single sided liquidity from the onJoin and onExit hook', async () => {})
+  it('totalUnclaimedFeesInNumeraire must be minted during onJoin or onExit', async () => {
```
Comment (ctverceles) [inline tests/local/pool.test.ts:419]: @0xAplki i think we also need to add a test here that expects a non 0 value after a swap happens and that LP tokens are minted after onJoin/onExit. so for example:
1. expect that `await fxPool.totalUnclaimedFeesInNumeraire()` returns an expected value (console log the value at this point before we do anything) 
2. trigger a swap
3. expect that `await fxPool.totalUnclaimedFeesInNumeraire()` returns the previous value + the new fees generated
4. trigger onJoinPool hook
5. expect the onJoinPool event emit
6. expect that `await fxPool.totalUnclaimedFeesInNumeraire()` is 0
7. expect that LP tokens were minted
8. trigger a swap
9. expect that `await fxPool.totalUnclaimedFeesInNumeraire()` returns the the new fees generated from the swap
10. trigger onExitPool hook
5. expect the onExitPool event emit
6. expect that `await fxPool.totalUnclaimedFeesInNumeraire()` is 0
11. expect that LP tokens were minted

**Review by ctverceles — CHANGES_REQUESTED** [review]
@0xAplki took a deeper look at the fee tests, i think we need to add in a non 0 test. Also put a suggestion for var naming but the fee recording logic looks good to me now thanks!

File: tests/local/pool.test.ts [hunk 8]
```diff
@@ -570,37 +410,132 @@ describe('FXPool', () => {
 
   // it('Previews swap caclculation from the onSwap hook using queryBatchSwap() ', async () => {})
```
Comment (ctverceles) [inline tests/local/pool.test.ts:411]: @0xAplki are we still gonna impl this or should we remove?

File: tests/local/pool.test.ts [hunk 9]
```diff
@@ -570,37 +410,132 @@ describe('FXPool', () => {
...
   // it('Previews swap caclculation when providing single sided liquidity from the onJoin and onExit hook', async () => {})
```
Comment (ctverceles) [inline tests/local/pool.test.ts:412]: @0xAplki are we still gonna impl this or should we remove?

File: tests/local/pool.test.ts [hunk 10]
```diff
@@ -570,37 +410,132 @@ describe('FXPool', () => {
...
+      fxPHPDecimals, // specify token in decimals
+      adminAddress,
+      adminAddress, // the account swapping should get the output tokens
+      fxPool,
+      testEnv,
+      log
+    )
+
+    const currentFeeBalanceAfterSwapBeforeDeposit = await fxPool.totalUnclaimedFeesInNumeraire()
+    expect(currentFeeBalanceAfterSwapBeforeDeposit).is.gt(previousFeeBalance)
```
Comment (ctverceles) [inline tests/local/pool.test.ts:436]: @0xAplki a bit of extra work - but can we show this to Khidir and see if the fee result makes sense to him? if so maybe we can have an expected fee balance that we calculate in parallel to the swap tx. then we compare it against the `currentFeeBalanceAfterSwapBeforeDeposit` and expect that its equal?

File: tests/local/pool.test.ts [hunk 11]
```diff
@@ -570,37 +410,132 @@ describe('FXPool', () => {
...
+    const lpTokensAfterDeposit = await fxPool.balanceOf(adminAddress)
+
+    expect(currentFeeBalanceAfterDeposit).is.equals(0)
+    expect(lpTokensAfterDeposit).to.equals(
+      lpTokensBeforeDeposit.add(currentFeeBalanceAfterSwapBeforeDeposit).add(depositDetails[0])
+    ) // including LP tokens after deposit
+  })
+
+  it('totalUnclaimedFeesInNumeraire must be minted during onExit', async () => {
+    const previousFeeBalance = await fxPool.totalUnclaimedFeesInNumeraire()
```
Comment (ctverceles) [inline tests/local/pool.test.ts:476]:
```suggestion
    const previousFeeBalance = await fxPool.totalUnclaimedFeesInNumeraire()
    expect(previousFeeBalance).is.equals(0)
```
@0xAplki just to be airtight?

File: tests/local/pool.test.ts [hunk 12]
```diff
@@ -570,37 +410,132 @@ describe('FXPool', () => {
...
     )
-    const amountIn1 = formatUnits(otherTokenIn, usdcDecimals)
+    // TODO: here
 
-    const sortedAmounts = sortDataLikeVault(sortedAddresses, fxPHPAddress, [
-      parseUnits(amountIn0),
-      parseUnits(amountIn1),
-    ])
+    const currentFeeBalanceAfterSwapBeforeWithdraw = await fxPool.totalUnclaimedFeesInNumeraire()
+    expect(currentFeeBalanceAfterSwapBeforeWithdraw).is.gt(previousFeeBalance)
```
Comment (ctverceles) [inline tests/local/pool.test.ts:499]: @0xAplki a bit of extra work - but can we show this to Khidir and see if the fee result makes sense to him? if so maybe we can have an expected fee balance that we calculate in parallel to the swap tx. then we compare it against the `currentFeeBalanceAfterSwapBeforeWithdraw ` and expect that its equal?

File: tests/local/assimilators.test.ts [hunk 13]
```diff
@@ -0,0 +1,626 @@
+import { expect } from 'chai'
```
Comment (ctverceles) [inline tests/local/assimilators.test.ts:1]: @0xAplki I guess this was just added in from the prior approved assimilator PR

File: contracts/core/FXSwaps.sol [hunk 14]
```diff
@@ -205,4 +214,13 @@ library FXSwaps {
...
         return Assimilators.viewNumeraireAmountAndBalance(_assim, _amt, address(curve.vault), curve.poolId);
     }
+
+    // internal function to avoid stack too deep
+    function _calculateFeeInNumeraire(int128 _outputNumeraireAmt, int128 _inputNumeraireAmt)
+        internal
+        pure
+        returns (int128 feeInNumeraire)
+    {
+        return _outputNumeraireAmt.sub(_inputNumeraireAmt);
```
Comment (ctverceles) [inline contracts/core/FXSwaps.sol:224]:
```suggestion
        return _inputNumeraireAmt.sub(_outputNumeraireAmt);
```
@0xAplki actually reading this again... shouldn't we do this the other way around like suggested here? cause `_inputNumeraireAmt` is the amount taken from user (before subtracting the swap fees). when we give the `_outputNumeraireAmt` back to the user it should be less cause lambda + epsilon was subtracted from the `_inputNumeraireAmt`.

File: contracts/core/FXSwaps.sol [hunk 15]
```diff
@@ -53,10 +54,14 @@ library FXSwaps {
             int128[] memory _oBals
         ) = viewOriginSwapData(curve, _o.ix, _t.ix, _originAmount, _o.addr);
 
+        int128 inputNumeraireAmount = _amt;
+
         _amt = CurveMath.calculateTrade(curve, _oGLiq, _nGLiq, _oBals, _nBals, _amt, _t.ix);
 
+        // negative for origin swap
+        accruedFees_ = -(_calculateFeeInNumeraire(_amt, inputNumeraireAmount));
```
Comment (ctverceles) [inline contracts/core/FXSwaps.sol:62]: @0xAplki can we run this by Khidir to see if he thinks this makes sense also?

File: contracts/core/FXSwaps.sol [hunk 16]
```diff
@@ -103,7 +110,9 @@ library FXSwaps {
         }
 
         _amt = _amt.us_mul(ONE + curve.epsilon);
+        accruedFees_ = _calculateFeeInNumeraire(_amt, inputNumeraireAmount);
```
Comment (ctverceles) [inline contracts/core/FXSwaps.sol:113]: @0xAplki can we run this by Khidir to see if he thinks this makes sense also? cause in `viewOriginSwap`, we're negating but here we're not. maybe khidir can just take a look at the math as well

File: contracts/FXPool.sol [hunk 17]
```diff
@@ -475,9 +485,22 @@ contract FXPool is IMinimalSwapInfoPool, BalancerPoolToken, Ownable, Storage, Re
         require(total + _amount < curve.cap, 'FXPool/amount-beyond-set-cap');
     }
 
-    /// @notice view the assimilator address for a derivative
-    /// @return assimilator_ the assimilator address
-    function assimilator(address _derivative) public view returns (address assimilator_) {
```
Comment (ctverceles) [inline contracts/FXPool.sol:480]: @0xAplki no need for this anymore?

**Review by ctverceles — CHANGES_REQUESTED** [review]
@0xAplki new test flow makes sense and works locally! 

few questions/suggestions below, though

File: tests/local/pool.test.ts (hunk 8 above)
Comment (0xAplki) [reply tests/local/pool.test.ts:411]: i actually ask james about this, but I think we have another way of doing things. If we dont want to leave comments, we can just note this on the ticket.

File: tests/local/pool.test.ts (hunk 9 above)
Comment (0xAplki) [reply tests/local/pool.test.ts:412]: same comment above

File: tests/local/pool.test.ts (hunk 10 above)
Comment (0xAplki) [reply tests/local/pool.test.ts:436]: sure!

File: tests/local/assimilators.test.ts (hunk 13 above)
Comment (0xAplki) [reply tests/local/assimilators.test.ts:1]: the whole thing? yup! but i changed the inputs to numeraire inputs

File: contracts/core/FXSwaps.sol (hunk 14 above)
Comment (0xAplki) [reply contracts/core/FXSwaps.sol:224]: afaik, fees are added on top of swap hence the current implementation

File: contracts/core/FXSwaps.sol [hunk 18]
```diff
@@ -53,10 +54,14 @@ library FXSwaps {
             int128[] memory _oBals
         ) = viewOriginSwapData(curve, _o.ix, _t.ix, _originAmount, _o.addr);
 
+        int128 inputNumeraireAmount = _amt;
+
         _amt = CurveMath.calculateTrade(curve, _oGLiq, _nGLiq, _oBals, _nBals, _amt, _t.ix);
 
+        // negative for origin swap
+        accruedFees_ = -(_calculateFeeInNumeraire(_amt, inputNumeraireAmount));
         _amt = _amt.us_mul(ONE - curve.epsilon);
```
Comment (ctverceles) [inline contracts/core/FXSwaps.sol:63]:
```suggestion
        _amt = _amt.us_mul(ONE - curve.epsilon);
        accruedFees_ = -(_calculateFeeInNumeraire(_amt, inputNumeraireAmount));
```
@0xAplki i think we'll also need to switch these 2 lines like this cause `_calculateFeeInNumeraire` won't include the epsilon fee in the minting

File: contracts/FXPool.sol (hunk 17 above)
Comment (0xAplki) [reply contracts/FXPool.sol:480]: its already in the storage plus we have an assilimiator factory. i don't see it used so far also.

File: contracts/core/FXSwaps.sol (hunk 16 above)
Comment (ctverceles) [reply contracts/core/FXSwaps.sol:113]: @0xAplki i see that

- in `viewOriginSwap()` its doing `ONE - curve.epsilon`
- and in `viewTargetSwap()` its doing `ONE + curve.epsilon`

I guess this is a note to run by Khidir when coming up w the expected values

File: contracts/FXPool.sol [hunk 19]
```diff
@@ -475,9 +487,24 @@ contract FXPool is IMinimalSwapInfoPool, BalancerPoolToken, Ownable, Storage, Re
...
+
+        totalUnclaimedFeesInNumeraire += feesToAdd;
+
+        emit FeesAccrued(feesToAdd);
+    }
+
+    function _mintProtocolFees() private {
+        require(collectorAddress != address(0), 'FXPool/fee-collector-not-set');
+
+        uint256 feesToCollect = totalUnclaimedFeesInNumeraire;
```
Comment (ctverceles) [inline contracts/FXPool.sol:504]:
```suggestion
        uint256 feesToCollect = totalUnclaimedFeesInNumeraire;
```
@0xAplki i just realized we don't have a var that records the % of `totalUnclaimedFeesInNumeraire` that we wanna give to the protocol. cause right now its minting LP tokens for total LP fees when we should only get a commission

File: tests/local/pool.test.ts [hunk 20]
```diff
@@ -372,25 +223,19 @@ describe('FXPool', () => {
...
     console.log('liquidity number', await ethers.utils.formatEther(liquidity))
     console.log('liquidity BigNumber', liquidity.toString())
-    //  expect(await ethers.utils.formatEther(liquidity)).to.be.equals(EXPECTED_LIQUIDITY)
-    // await expect(liquidity, 'unexpected liquidity() result')
-    //   .to.be.greaterThan(BigNumber.from(10000))
-    //   .lessThan(BigNumber.from(10001))
+    expect(liquidity).to.not.equals(0)
   })
 
   it('originSwap: User batch swaps token A (USDC) for token B (fxPHP) calling the vault and triggering the onSwap hook', async () => {
```
Comment (ctverceles) [inline tests/local/pool.test.ts:232]: @0xAplki i think its time to expect actual values on lines 455 to 457

File: tests/local/pool.test.ts [hunk 21]
```diff
@@ -461,9 +306,9 @@ describe('FXPool', () => {
   it('originSwap: User batch swaps token A (fxPHP) for token B (USDC) calling the vault and triggering the onSwap hook', async () => {
```
Comment (ctverceles) [inline tests/local/pool.test.ts:306]: @0xAplki i think its time to expect actual values on lines 410 to 412

File: tests/local/pool.test.ts [hunk 22]
```diff
@@ -570,37 +415,157 @@ describe('FXPool', () => {
 
   // it('Previews swap caclculation from the onSwap hook using queryBatchSwap() ', async () => {})
   // it('Previews swap caclculation when providing single sided liquidity from the onJoin and onExit hook', async () => {})
```
Comment (ctverceles) [inline tests/local/pool.test.ts:417]:
```suggestion
```
remove comments, note on ticket instead

**Review by ctverceles — CHANGES_REQUESTED** [review]
Small notes on the tests @0xAplki

File: contracts/FXPool.sol (hunk 19 above)
Comment (0xAplki) [reply contracts/FXPool.sol:514]: made a different implementation for dynamic protocol fee percentage setting

File: tests/local/pool.test.ts (hunk 20 above)
Comment (0xAplki) [reply tests/local/pool.test.ts:233]: Filed: https://halodao.atlassian.net/browse/HDEV-658

File: tests/local/pool.test.ts (hunk 21 above)
Comment (0xAplki) [reply tests/local/pool.test.ts:308]: filed: https://halodao.atlassian.net/browse/HDEV-658

**Comment by 0xAplki** [comment]
Merging this from our conversation @ctverceles

**Pull request merged**